### DISCUSSION TITLE
Is it possible to modify the LogEventData object?

### DISCUSSION THREAD

**Discussion by schallm** [datalust/seq-tickets#1848] [Help] [answered]
We are trying to move to `Seq` and want to use structured logging.  Most of our code still uses `log4net` but we are moving to `nLog`.  We have a few log statements that we would like to aggregate some properties.  For example, we log our database calls with a time taken appended at the end of the message.  Once we migrate to `nLog` we will log the time "structurally" so it would just be another property.  However, that will take some time.

Is it possible to do this as part of ingest?  I tried to right the following `SeqApp` that inspects the `logger` property, pulls the values out of the `RenderedMessage`, and adds matches to the parameters.  While the App installs fine and seems to process the messages sent to it, I assume I'm modifying the `Properties` collection too late? 

Is something like this supported or am I out of luck?

```cs
using Seq.Apps;
using Seq.Apps.LogEvents;
using System.Text.RegularExpressions;

namespace Seq.App.FauxStructuredLogging;

[SeqApp("Faux Structured Logging")]
public class FauxStructuredLoggingApp : SeqApp, ISubscribeTo<LogEventData> {

    private static Regex _queryResultRegex = new(@"/\*time:(?<Duration>[\.\d]+)ms\*/$", RegexOptions.Compiled);

    public void On(Event<LogEventData> evt) {
        var logger = evt.Data.Properties.GetValueOrDefault("Logger") as string;
        if (!string.IsNullOrWhiteSpace(logger)) {
            if (logger.StartsWith("Foo.Bar.")) {
                foreach (Match match in _queryResultRegex.Matches(evt.Data.RenderedMessage)) {
                    if (match.Success) {
                        var newProperties = new Dictionary<string, object>(evt.Data.Properties);
                        // don't use 0th match as that is the full match
                        for (var i = 1; i < match.Groups.Count; i++) {
                            newProperties[match.Groups[i].Name] = decimal.Parse(match.Groups[i].Value);
                        }
                        evt.Data.Properties = newProperties;
                    }
                }
            }
        }
    }

}
```

**Comment by nblumhardt** [comment, marked as answer]
Hi Mike, thanks for dropping us a line.

This needs to be done on the client side at present - either via modified log4net integration (you can find Seq.Client.Log4Net at https://github.com/datalust/seq-client-log4net which may suggest some possibilities), or by pushing logs through a preprocessor such as Fluent Bit (we have some discussion of that [on our blog](https://blog.datalust.co/tailing-a-log-file-or-folder-with-fluent-bit-and-seq/) - this post reads from log files, but Fluent Bit can accept logs via the network, too).

Let me know if any of the above helps,

Nick

**Reply by schallm** [reply]
We are going to modify how the log4net appender sends the information.  Thanks for the excellent support!
